### PR TITLE
[INFRA/RULES] Infra rules baseclient functions added

### PR DESCRIPTION
**Pull request by Neelesh99**
# Summary

Added a couple of functions to the basclient that make use of the newly developed toolkit in #275. These include compliance checkers for Tax, Sanctions and Allocations as well as Compliant Value recommendation. 

## Additional Information

Once the new (and exhaustive) set of rules are added, I can extend these baseclient functions to cover them, for now this might provide teams with a  starting point. I've also added a LocalVariableCache to baseclient which stores the required variables that are associated with rules.

## Future tools
Since we don't have feature freeze till the 6th, feel free to write on this PR or PM me what extra functionality you'd like added to baseclient since I recognise this is not all that people wanted, just what I thought might be useful off the top of my head.

### CODE REVIEW

**Comment by QFSW** [comment]
# ⚠️Warning! This diff contains changes in protected folders

**Comment by Neelesh99** [comment]
The high LOC is because this PR is linearly dependant on #275, once that goes in the PR should be like 150 - 200 LOC

File: internal/common/baseclient/iigo.go [hunk 1]
```diff
@@ -17,8 +17,10 @@ func (c *BaseClient) CommonPoolResourceRequest() shared.Resources {
 // the President to help work out how many resources to allocate each island.
 // OPTIONAL : as is, this function will always report island resources accurately
 func (c *BaseClient) ResourceReport() shared.ResourcesReport {
+	amountReported := c.ServerReadHandle.GetGameState().ClientInfo.Resources
+	c.LocalVariableCache[rules.IslandReportedResources] = rules.MakeVariableValuePair(rules.IslandReportedResources, []float64{float64(amountReported)})
 	return shared.ResourcesReport{
-		ReportedAmount: c.ServerReadHandle.GetGameState().ClientInfo.Resources,
+		ReportedAmount: amountReported,
```
Comment (Eirikalb) [inline internal/common/baseclient/iigo.go:23]: Nice refactor

File: internal/common/baseclient/iigo.go [hunk 2]
```diff
@@ -56,14 +58,46 @@ func (c *BaseClient) GetClientSpeakerPointer() roles.Speaker {
...
+	isCompliant := c.CheckCompliance(rules.IslandTaxContribution)
+	if isCompliant {
+		// TODO: with this compliance check, agents can see whether they'd like to continue returning this value
+		valToBeReturned = 0
+	}
+	// Use the toolkit to recommend a value
+	newVal, success := c.GetRecommendation(rules.IslandTaxContribution)
+	if success {
+		// TODO: Choose whether to use this compliant value
+		valToBeReturned = shared.Resources(newVal.Values[0])
```
Comment (Eirikalb) [inline internal/common/baseclient/iigo.go]: What is nevVal.Values[0]? could the 0 be encoded as an enum maybe for clarity?

Comment (Eirikalb) [reply internal/common/baseclient/iigo.go]: Same on line 84 and 98

File: internal/common/rules/clienttools_test.go [hunk 3]
```diff
@@ -0,0 +1,505 @@
+package rules
```
Comment (Eirikalb) [inline internal/common/rules/clienttools_test.go:1]: LOVE THIS FILE! We love tests!

File: internal/common/rules/clienttools.go [hunk 4]
```diff
@@ -0,0 +1,174 @@
+package rules
```
Comment (Eirikalb) [inline internal/common/rules/clienttools.go:1]: These are for the clients to use right? Great addition!

File: internal/clients/team3/client.go [hunk 5]
```diff
@@ -18,9 +18,9 @@ type client struct {
 }
```
Comment (Eirikalb) [inline internal/clients/team3/client.go:18]: I suppose this is due to broken backwards compatibility?

**Review by Eirikalb — APPROVED** [review]
LGTM, I will pre-approve, but you need to fix linting errors and would be nice to fix that nevVal.Values(0) thing before merge. I dont get github notifications. Overall stuff looks reasonable, thanks for adding to the BaseClient, makes me very happy <3

File: internal/common/baseclient/iigo.go (hunk 2 above)
Comment (Neelesh99) [reply internal/common/baseclient/iigo.go]: Yeah good point, will do

File: internal/clients/team3/client.go (hunk 5 above)
Comment (Neelesh99) [reply internal/clients/team3/client.go:18]: Yup, function name change

File: internal/common/baseclient/iigo.go (hunk 2 above)
Comment (Neelesh99) [reply internal/common/baseclient/iigo.go]: Done

File: internal/common/rules/clienttools.go (hunk 4 above)
Comment (Neelesh99) [reply internal/common/rules/clienttools.go:1]: Yup

**Comment by Eirikalb** [comment]
Did you fix the linting errors? The build failed....

**Pull request merged**